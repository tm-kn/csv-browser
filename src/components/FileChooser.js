import React, { Component, PropTypes } from 'react';

export default class FileChooser extends Component {
  static propTypes = {
    accept: PropTypes.string,
    onChangeFile: PropTypes.func.isRequired
  };

  constructor() {
    super();

    this.state = {
      changingFile: undefined
    };

    this.handleChangeFile = this.handleChangeFile.bind(this);
    this.handleSubmitFile = this.handleSubmitFile.bind(this);
  }

  render() {
    return (
      <div>
        <h1>Select file</h1>
        <p>Select the file you want to open.</p>
        <form onSubmit={this.handleSubmitFile}>
          <input
            accept={this.props.accept}
            name="csvFile"
            onChange={this.handleChangeFile}
            type="file"
          />
          <input
            className="button"
            disabled={!this.state.changingFile}
            type="submit"
            value="Load to the browser"
          />
        </form>
      </div>
    );
  }

  handleChangeFile(event) {
    const target = event.target;

    this.setState(() => ({
      changingFile: target.files[0]
    }));
  }

  handleSubmitFile(event) {
    event.preventDefault();

    if (!this.state.changingFile) {
      alert('Please choose a file first');
      return;
    }

    this.props.onChangeFile(this.state.changingFile);

    this.setState(state => ({
      changingFile: undefined
    }));
  }
}
