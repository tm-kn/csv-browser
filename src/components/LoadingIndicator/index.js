import React, { Component } from 'react';

import './loading-indicator.css';
import image from './loading-image.svg';

export default class LoadingIndicator extends Component {
  render() {
    return (
      <div className="Loading-container">
        <img
          alt="Loading indicator"
          src={image}
          className="Loading-indicator"
        />
        <h2>Loading...</h2>
      </div>
    );
  }
}
