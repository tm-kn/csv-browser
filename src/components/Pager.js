import React, { Component, PropTypes } from 'react';

export default class Pager extends Component {
  static defaultProps = {
    offsetValues: [15, 30, 60, 90]
  };

  static propTypes = {
    numberOfRecords: PropTypes.number.isRequired,
    offset: PropTypes.number.isRequired,
    offsetValues: PropTypes.arrayOf(PropTypes.number),
    onChangePage: PropTypes.func.isRequired,
    onChangeOffset: PropTypes.func.isRequired,
    page: PropTypes.number.isRequired
  };

  get numberOfPages() {
    const numberOfPages = Math.ceil(this.props.numberOfRecords / this.props.offset);

    return numberOfPages || 1;
  }

  constructor() {
    super();

    this.state = {
      changingPageNumber: ''
    };

    this.handleGoToPage = this.handleGoToPage.bind(this);
    this.handleNextPage = this.handleNextPage.bind(this);
    this.handleChangeOffset = this.handleChangeOffset.bind(this);
    this.handlePreviousPage = this.handlePreviousPage.bind(this);
    this.handleSubmitPageNumber = this.handleSubmitPageNumber.bind(this);
    this.previousPageExists = this.previousPageExists.bind(this);
    this.nextPageExists = this.nextPageExists.bind(this);
  }

  render() {
    return (
      <div>
        <button
          className="button"
          disabled={this.props.page === 1}
          onClick={() => this.handleGoToPage(1)}
        >
          First page
        </button>
        <button
          className="button"
          disabled={!this.previousPageExists()}
          onClick={this.handlePreviousPage}
        >
          Previous page
        </button>
        <span>
          {this.props.page} / {this.numberOfPages}
        </span>
        <button
          className="button"
          disabled={!this.nextPageExists()}
          onClick={this.handleNextPage}
        >
          Next page
        </button>
        <button
          className="button"
          disabled={this.state.page === this.numberOfPages}
          onClick={() => this.handleGoToPage(this.numberOfPages)}
        >
          Last page
        </button>
        <form onSubmit={this.handleSubmitPageNumber}>
          <input
            type="number"
            name="page"
            placeholder="Page number..."
            onChange={(event) =>  {
              const target = event.target;
              this.setState(() => ({
                changingPageNumber: target.value
              }));
            }}
            value={this.state.changingPageNumber}
          />
          <input
            className="button"
            type="submit"
            name="submitPageNumber"
            value="Go"
          />
        </form>
        <div>
          <p>Records found: {this.props.numberOfRecords}</p>
          <ul className="horizontal">
            <li>
              <span>Display:</span>
            </li>
            {this.props.offsetValues.map((item, key) => (
              <li key={key}>
                <button
                  disabled={this.props.offset === item}
                  className="button"
                  onClick={() => this.handleChangeOffset(item)}
                >
                  {item} records
                </button>
              </li>
            ))}
          </ul>
        </div>
      </div>
    );
  }

  handleGoToPage(page) {
    let newPage;

    if (newPage > this.numberOfPages) {
      newPage = this.numberOfPages;
    } else if (newPage < 1) {
      newPage = 1;
    } else {
      newPage = page;
    }

    this.props.onChangePage(newPage);
  }

  handleNextPage() {
    this.handleGoToPage(this.props.page + 1);
  }

  handleChangeOffset(offset) {
    const currentStartIndex = ((this.props.page - 1) * (this.props.offset));
    const newPage = Math.floor(currentStartIndex / offset) + 1;
    
    this.props.onChangeOffset(offset);
    this.handleGoToPage(newPage);
  }

  handlePreviousPage() {
    this.handleGoToPage(this.props.page - 1);
  }

  nextPageExists() {
    return (this.props.page < this.numberOfPages);
  }

  previousPageExists() {
    return (this.props.page > 1);
  }

  handleSubmitPageNumber(event) {
    event.preventDefault();

    if (
      this.state.changingPageNumber > this.numberOfPages
      || this.state.changingPageNumber <= 0
    ) {
      alert(`${this.state.changingPageNumber} is not a valid page number.`);
      return;
    }

    const pageNumber = parseInt(this.state.changingPageNumber, 10);
    this.setState(() => ({ changingPageNumber: '' }));
    this.handleGoToPage(parseInt(pageNumber, 10));
  }
}