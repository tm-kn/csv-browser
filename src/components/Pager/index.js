import { Component, PropTypes } from 'react';

import template from './Pager.jsx';

export default class Pager extends Component {
  static defaultProps = {
    offsetValues: [15, 30, 60, 90]
  };

  static propTypes = {
    numberOfRecords: PropTypes.number.isRequired,
    offset: PropTypes.number.isRequired,
    offsetValues: PropTypes.arrayOf(PropTypes.number),
    onChangePage: PropTypes.func.isRequired,
    onChangeOffset: PropTypes.func.isRequired,
    page: PropTypes.number.isRequired
  };

  get numberOfPages() {
    const numberOfPages = Math.ceil(this.props.numberOfRecords / this.props.offset);

    return numberOfPages || 1;
  }

  constructor() {
    super();

    this.state = {
      changingPageNumber: ''
    };

    this.handleGoToPage = this.handleGoToPage.bind(this);
    this.handleNextPage = this.handleNextPage.bind(this);
    this.handleChangeOffset = this.handleChangeOffset.bind(this);
    this.handlePreviousPage = this.handlePreviousPage.bind(this);
    this.handleSubmitPageNumber = this.handleSubmitPageNumber.bind(this);
    this.previousPageExists = this.previousPageExists.bind(this);
    this.nextPageExists = this.nextPageExists.bind(this);
  }

  render() {
    return template.call(this);
  }

  handleGoToPage(page) {
    let newPage;

    if (newPage > this.numberOfPages) {
      newPage = this.numberOfPages;
    } else if (newPage < 1) {
      newPage = 1;
    } else {
      newPage = page;
    }

    this.props.onChangePage(newPage);
  }

  handleNextPage() {
    this.handleGoToPage(this.props.page + 1);
  }

  handleChangeOffset(offset) {
    const currentStartIndex = ((this.props.page - 1) * (this.props.offset));
    const newPage = Math.floor(currentStartIndex / offset) + 1;
    
    this.props.onChangeOffset(offset);
    this.handleGoToPage(newPage);
  }

  handlePreviousPage() {
    this.handleGoToPage(this.props.page - 1);
  }

  nextPageExists() {
    return (this.props.page < this.numberOfPages);
  }

  previousPageExists() {
    return (this.props.page > 1);
  }

  handleSubmitPageNumber(event) {
    event.preventDefault();

    if (
      this.state.changingPageNumber > this.numberOfPages
      || this.state.changingPageNumber <= 0
    ) {
      alert(`${this.state.changingPageNumber} is not a valid page number.`);
      return;
    }

    const pageNumber = parseInt(this.state.changingPageNumber, 10);
    this.setState(() => ({ changingPageNumber: '' }));
    this.handleGoToPage(parseInt(pageNumber, 10));
  }
}
