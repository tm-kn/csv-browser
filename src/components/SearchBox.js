import React, { Component, PropTypes } from 'react';

export default class SearchBox extends Component {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    value: PropTypes.string.isRequired
  };

  constructor() {
    super();

    this.state = {
      changingSearch: ''
    };

    this.handleSubmitSearch = this.handleSubmitSearch.bind(this);
  }

  render() {
    if (this.props.value) {
      return (
        <p>Searching by: {this.props.value}</p>
      );
    }

    return (
      <form onSubmit={this.handleSubmitSearch}>
        <input
          type="text"
          name="search"
          placeholder="Search..."
          onChange={(event) => {
            const target = event.target;

            this.setState(() => ({
              changingSearch: target.value
            }));
          }}
          value={this.state.changingSearch}
        />
        <input
          type="submit"
          className="button"
          name="submitSearch"
          value="Search"
        />
      </form>
    );
  }

  handleSubmitSearch(event) {
    event.preventDefault();

    if (!this.state.changingSearch) {
      alert('Search string cannot be empty');
      return;
    }

    this.props.onChange(this.state.changingSearch);

    this.setState(state => ({
      changingSearch: ''
    }));

  }
}
