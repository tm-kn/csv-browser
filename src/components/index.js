import FileChooser from './FileChooser';
import LoadingIndicator from './LoadingIndicator';
import Pager from './Pager';
import SearchBox from './SearchBox';

export {
    FileChooser,
    LoadingIndicator,
    Pager,
    SearchBox
};
