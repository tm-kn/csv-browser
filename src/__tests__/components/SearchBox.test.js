import React from 'react';
import renderer from 'react-test-renderer';

import { SearchBox } from 'components';

it('renders without crashing', () => {
  const tree = renderer.create(<SearchBox value={''} onChange={() => {}}/>);
  
  expect(tree.toJSON()).toMatchSnapshot();
});
