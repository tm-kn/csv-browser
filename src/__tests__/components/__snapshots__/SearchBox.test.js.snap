exports[`test renders without crashing 1`] = `
<form
  onSubmit={[Function]}>
  <input
    name="search"
    onChange={[Function]}
    placeholder="Search..."
    type="text"
    value="" />
  <input
    className="button"
    name="submitSearch"
    type="submit"
    value="Search" />
</form>
`;
